platform/src/lib.rs

```diff
@@ -545,6 +590,66 @@ pub extern "C" fn roc_fx_fileReadBytes(
     }
 }
 
+#[no_mangle]
+pub extern "C" fn roc_fx_fileReader(roc_path: &RocList<u8>) -> RocResult<u64, roc_app::ReadErr> {
+    match File::open(path_from_roc_path(roc_path)) {
+        Ok(file) => READERS.with(|reader_thread_local| {
+            let mut readers = reader_thread_local.borrow_mut();
+
+            let new_index = readers.len();
+
+            let buf_reader = BufReader::new(file);
+
+            readers.push(Rc::new(RefCell::new(Box::new(buf_reader))));
+
+            RocResult::ok(new_index as u64)
+        }),
+        Err(err) => RocResult::err(toRocReadError(err)),
+    }
+}
+
+#[no_mangle]
+pub extern "C" fn roc_fx_fileReadLine(readerIndex: u64) -> RocResult<RocList<u8>, RocStr> {
+    READERS.with(|reader_thread_local| {
+        let readers = reader_thread_local.borrow_mut();
+
+        let reader_ref_cell = readers.get(readerIndex as usize).expect("readerIndex was larger than number of readers in Vec. We expect this can only happen due to an implementation error in basic-cli.");
+
+        let mut string_buffer = String::new();
+
+        let mut file_buf_reader = reader_ref_cell.borrow_mut();
+
+        match file_buf_reader.read_line(&mut string_buffer) {
+            Ok(bytes_read) => {
+                if bytes_read == 0 {
+                    // return empty list when no bytes read, e.g. End Of File
+                    return RocResult::ok(RocList::empty());
+                }
+
+                RocResult::ok(RocList::from(string_buffer.as_bytes()))
+            }
+            Err(err) => RocResult::err(err.to_string().as_str().into()),
+        }
+    })
+}
+
+#[no_mangle]
+pub extern "C" fn roc_fx_closeFile(readerIndex: u64) {
+    READERS.with(|reader_thread_local| {
+        let readers = reader_thread_local.borrow_mut();
+
+        let reader_ref_cell = readers.get(readerIndex as usize).expect("readerIndex was larger than number of readers in Vec. We expect this can only happen due to an implementation error in basic-cli.");
+
+        let dummy_reader = BufReader::new(std::io::Cursor::new(Vec::new()));
+
+        // we drop the BufReader<File> to manually close the file
+        // by replacing it with a dummy reader that will return EOF
+        // for any future read attempts.
+        *reader_ref_cell.borrow_mut() = Box::new(dummy_reader);
+
+    })
+}
```

Comment:
@smores56 what do you think of this? 

I'm a rust noob, but this seems to be working ok. (segfault on linux aside -- which seems unrelated)

Reply:
My instinct is to say we should just remove it from the READERS list, but that would break the indices of the other readers. So unless we change this to a `HashMap`, or a ring-buffer of some sort, the current way this works is fine. It will just grow over time, which is sub-optimal.

Reply:
Done, switch to a `HashMap` which is much simpler.